src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,233 @@
+require "http"
 require "./helpers"
+require "../config"
+require "./file_watcher"
+require "../../exceptions/exception_page"
 
-module Sentry
-  class ProcessRunner
-    property processes = [] of Process
-    property process_name : String
-    property files = [] of String
-    @logger : Amber::Environment::Logger
-    FILE_TIMESTAMPS = {} of String => String
-
-    def initialize(
-      @process_name : String,
-      @build_command : String,
-      @run_command : String,
-      @build_args : Array(String) = [] of String,
-      @run_args : Array(String) = [] of String,
-      files = [] of String,
-      @logger = Amber::CLI.logger
-    )
-      @files = files
-      @npm_process = false
-      @app_running = false
+module Amber::CLI::Helpers
+  include Environment
+
+  struct ProcessRunner
+    PROCESSES = [] of {Process, String}
+
+    @host : String
+    @port : Int32
+    @file_watcher : FileWatcher
+
+    def initialize
+      @watch_running = false
+      @wait_build = Channel(Bool).new
+      @server_files_changed = false
+      @notify_counter = 0
+      @notify_counter_channel = Channel(Int32).new
+      @notify_channel = Channel(Nil).new
+      @host = Helpers.settings.host
+      @port = Helpers.settings.port
+      @file_watcher = FileWatcher.new
+
+      at_exit do
+        kill_processes
+      end
+
+      Signal::INT.trap do
+        Signal::INT.reset
+        exit
+      end
     end
 
     def run
-      loop do
-        scan_files
-        sleep 1
-      end
-    end
-
-    # Compiles and starts the application
-    def start_app
-      build_result = build_app_process
-      if build_result.is_a? Process::Status
-        if build_result.success?
-          stop_all_processes
-          create_all_processes
-          @app_running = true
-        elsif !@app_running
-          log "Compile time errors detected. Shutting down..."
-          exit 1
+      if watch_config = CLI.config.watch
```

Comment:
This should be moved to initializer as a guard clause. watch
```crystal
def initialize
  raise AMBER_YAML_ERROR unless CLI.config.amber_yml_exists?
  @amber_yaml =  CLI.config.amber_yaml
end
```

Actually the loading and check of this should be in Amber::CLI module and not here. 

And have an instance of ProcessRunner in Amber::CLI 

```crystal
module Amber::CLI
  def self.process_runner
    ProcessRunner.new if amber_yml_loaded?
  end

  def amber_yaml_loaded?
     raise AMBER_YAML_ERROR unless CLI.config.amber_yml_exists?
  end
end

```

Reply:
Nice suggestion, let me apply this refactoring as well :+1: